pkg/handlers/generic/lifecycle/csi/handler.go

```diff
@@ -74,68 +74,117 @@ func (c *CSIHandler) AfterControlPlaneInitialized(
 		)
 		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
 		resp.SetMessage(
-			fmt.Sprintf("failed to read CSI provider from cluster definition: %v",
+			fmt.Sprintf("failed to read the CSI variable from the cluster: %v",
 				err,
 			),
 		)
 		return
 	}
-	if len(csiProviders.Providers) == 1 &&
-		len(csiProviders.Providers[0].StorageClassConfig) == 1 &&
-		csiProviders.DefaultStorage == nil {
-		csiProviders.DefaultStorage = &v1alpha1.DefaultStorage{
-			ProviderName:           csiProviders.Providers[0].Name,
-			StorageClassConfigName: csiProviders.Providers[0].StorageClassConfig[0].Name,
+
+	// This is defensive, because the API validation requires at least one provider.
+	if len(csi.Providers) == 0 {
+		log.Error(
+			err,
+			"The list of CSI providers must include at least one provider.",
+		)
+		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		resp.SetMessage("the list of CSI providers must include at least one provider")
+		return
+	}
+
+	storageClassConfigsByProviderName := map[string][]v1alpha1.StorageClassConfig{}
+	for _, provider := range csi.Providers {
+		storageClassConfigsByProviderName[provider.Name] = provider.StorageClassConfig
+	}
+	if configs, ok := storageClassConfigsByProviderName[csi.DefaultStorage.ProviderName]; !ok {
+		log.Error(
+			err,
+			"The default Storage provider name must be the name of a chosen default provider.",
+		)
+		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		resp.SetMessage("The default Storage provider name must be the name of a chosen provider")
+		return
+	} else {
+		defaultStorageClassConfigNameInProvider := false
+		for _, config := range configs {
+			if csi.DefaultStorage.StorageClassConfigName == config.Name {
+				defaultStorageClassConfigNameInProvider = true
+				break
+			}
+		}
+		if !defaultStorageClassConfigNameInProvider {
+			log.Error(
+				err,
+				"The default Storage StrorageClassConfig name must be the name of a StrorageClassConfig of the default provider.",
+			)
+			resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+			resp.SetMessage(
+				"The default Storage StrorageClassConfig name must be the name of a StrorageClassConfig of the default provider.",
+			)
+			return
 		}
 	}
 
-	// There's a 1:N mapping of infra to CSI providers. The user chooses the provider.
-	for _, provider := range csiProviders.Providers {
+	// There's a 1:N mapping of infra to CSI providers. The user chooses the providers.
+	for _, provider := range csi.Providers {
+		// This is defensive, because the API validation requires at least provider.
+		if len(provider.StorageClassConfig) == 0 {
+			log.Error(
+				err,
+				"The CSI provider must configure at least one storage class.",
+				"name",
+				provider.Name,
+			)
+			resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+			resp.SetMessage(
+				fmt.Sprintf("the CSI provider %q must configure at least one storage class",
+					provider.Name),
+			)
+			return
+		}
```

Comment:
I think we can delegate to the client (CLI/UI). The client will create at least one storage class, so the user won't need to do additional work.

Is there a good reason to not have a storage class? If there is, then we can only require a storage class to satisfy a reference in DefaultStorage.

Reply:
Sorry I might be misreading this, but it looks like we need to have a storage class for every CSI provider. I have a stronger disagreement with that, because just deploying a CSI provider seems like something a user could want to do.

Reply:
That's a good point!

How about we require only the storage class that is referenced by `DefaultStorage.StorageClassConfigName`? Since we require `DefaultStorage`, we create at least one storage class on the cluster, and make it the default one. Additional storage classes are optional.

Reply:
Done in https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/603/commits/e53524694936c0d242e3cb06c6164b4eef84dbd3